### PR TITLE
45736 Build and populate the year and week dropdowns

### PR DESCRIPTION
**Pull request by Shirin14926**
Code changes for Year Week dynamic dropdown
Unit Test cases , Automated test cases

_NOTE_: **Dynamic dropdown(PBI 45736) changes are incorporated in PBI 46346- Reskin/Uplift Weekly NM Content as UI content changes are closely intact and looking at dependency of this PBI in reskin, we have merged the change and only one PR will be raised for it**

saving this PR as draft

### CODE REVIEW

**Comment by barrie-cooper** [comment]
Sorry @Shirin14926, conflicts need resolving please.

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs [hunk 1]
```diff
@@ -49,55 +49,51 @@ public async Task<List<ShowFilesResponseModel>> GetWeeklyBatchFiles(int year, in
...
+                                {
+                                    string[] yearWeek = yw.Split('/');
+                                    yearWeekModelList.Add(new YearWeekModel { Year = yearWeek[0].Trim(), Week = yearWeek[1].Trim() });
+                                }
+                                _logger.LogInformation(EventIds.GetSearchAttributeRequestDataFound.ToEventId(), "Request Search Attribute Year and week data recieved successfully from File Share Service for BatchSearchAttribute with _X-Correlation-ID:{correlationId}" , correlationId);
+                            }
+                            else
+                            {
+                                _logger.LogInformation(EventIds.GetSearchAttributeRequestDataNotFound.ToEventId(), "No Data recieved from File Share Service for Request Search Attribute Year and week for _X-Correlation-ID:{correlationId}", correlationId);
+                            }
```
Comment (vijendr13545) [inline UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs:85]: Please break the outer for loop if `YEAR / WEEK` is found.

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/wwwroot/js/Custom/NoticesToMariners.js [hunk 2]
```diff
@@ -1,62 +1,90 @@
...
+
+        if (onload) {
+            var selectedweek = weekdata[weekdata.length - 1];
+            $('#ddlWeeks').val(selectedweek);
+            ShowWeeklyFilesAsync();
+            onload = false;
+        }
+        else {
+            $('#ddlWeeks').val('--Please select week--');
+        }
```
Comment (vijendr13545) [inline UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/wwwroot/js/Custom/NoticesToMariners.js:87]: We need to revisit it to remove Jquery dependency. I am happy if that is done as a separate PBI/task.

**Review by vijendr13545 — COMMENTED** [review]
Looks good once all review comments are fixed. Thanks.

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs (hunk 1 above)
Comment (Shirin14926) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs:89]: done

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/Models/NoticesToMariners/YearWeekModel.cs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+﻿namespace UKHO.MaritimeSafetyInformation.Common.Models.NoticesToMariners
+{
+    public class YearWeekModel
+    {
+        public string Year { get; set; }
+        public string Week { get; set; }
```
[the remark this answers is not in the document]
Comment (Shirin14926) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/Models/NoticesToMariners/YearWeekModel.cs]: Changed to int. Thanks

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs [hunk 4]
```diff
@@ -49,55 +49,54 @@ public async Task<List<ShowFilesResponseModel>> GetWeeklyBatchFiles(int year, in
...
-                lastdate = new DateTime(year, DateTime.Now.Month, DateTime.Now.Day);
-            }
-            else
-            {
-                lastdate = new DateTime(year, 12, 31);
+                if (searchAttributes.Data != null)
+                {
+                    for (int i = 0; i < searchAttributes.Data.BatchAttributes.Count; i++)
+                    {
+                        if (searchAttributes.Data.BatchAttributes[i].Key.Trim() == "YEAR/WEEK")
```
[the remark this answers is not in the document]
Comment (Shirin14926) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs]: Done

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/wwwroot/js/Custom/NoticesToMariners.js (hunk 2 above)
Comment (Shirin14926) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/wwwroot/js/Custom/NoticesToMariners.js:87]: As of Now, I have kept same jquery logic, will take it as a separate task.

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs (hunk 1 above)
Comment (vijendr13545) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs:89]: Please can you use  class level constant for it.

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs [hunk 5]
```diff
@@ -49,55 +49,51 @@ public async Task<List<ShowFilesResponseModel>> GetWeeklyBatchFiles(int year, in
...
-                lastdate = new DateTime(year, DateTime.Now.Month, DateTime.Now.Day);
-            }
-            else
-            {
-                lastdate = new DateTime(year, 12, 31);
+                if (searchAttributes.Data != null)
+                {
+                    for (int i = 0; i < searchAttributes.Data.BatchAttributes.Count; i++)
+                    {
+                        if (searchAttributes.Data.BatchAttributes[i].Key == "YEAR / WEEK")
```
[the remark this answers is not in the document]
Comment (vijendr13545) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs]: Please can you use class level constant for it

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs [hunk 6]
```diff
@@ -49,55 +49,54 @@ public async Task<List<ShowFilesResponseModel>> GetWeeklyBatchFiles(int year, in
...
+                        {
+                            List<string> yearWeekList = searchAttributes.Data.BatchAttributes[i].Values;
+                            if (yearWeekList != null && yearWeekList.Count != 0)
+                            {
+                                foreach (string yw in yearWeekList)
+                                {
+                                    string[] yearWeek = yw.Contains('/') ? yw.Split('/') : null;
+                                    if (yearWeek != null)
+                                    {
+                                        yearWeekModelList.Add(new YearWeekModel { Year = yearWeek[0].Trim(), Week = yearWeek[1].Trim() });
```
[the remark this answers is not in the document]
Comment (vijendr13545) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs]: We must check that array should have a length at-least 2 before accessing value at index 1. Can you correct the code accordingly.

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Shirin14926) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs]: defined as constant

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Shirin14926) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Web/Services/NMDataService.cs]: did changes as suggested, Thanks

**Comment by usha13529** [comment]
NOTE: Dynamic dropdown(PBI 45736) changes are incorporated in PBI 46346- Reskin/Uplift Weekly NM Content as UI content changes are closely intact and looking at dependency of this PBI in reskin, PR23 contains these changes, so this PR is not required hence closing this PR.

**Pull request closed without merging**